### ISSUE TITLE
Any Plans to Migrate to Scintilla v4?

### ISSUE THREAD

**Issue by tajmone** [fantaisie-software/purebasic#59]
Since this might be critical for anyone working on Scintilla code in the PB IDE, what are the plans regarding PureBasic migrating to Scintilla v4?

Version 4 has now become the official major release since quite a while, whereas PB still adopts a rather old v3.x release.

From the Scintilla documentation, it's clear that migrating to v4 would break backward compatibility in many places, so it's a big change indeed.

- What are the plans and ETA in this regard?
- Is PB going to abruptly switch from 3 to 4, or is there going to be a "grace period" during which end users can choose whether to use v3.x or v4.x?
- When the Scintilla v4 PureLib is released, does the PB IDE need to upgrade to it, or could it just keep using the old version until the code has been adapted?

A grace period is indeed desirable, since v4 breaks backward compatibility. But since Scintilla development is now focused on v4, it will eventually make no longer sense to keep supporting it — unless keeping the dual-choice is viable, after all once the Scintilla 3 PureLib gets updated to the latest v3.x, it won't need any further updates either.

Currently, it seems that a [LongTerm3] branch is still being maintained for old projects that depend on v3:

> the LongTerm3 branch of Scintilla […] avoids depending on C++14 or later features. Compilers as old as GCC 4.8, MSVC 2015 (theoretically), and Clang 3.4 are supported.

but I guess this will eventually come to an end.

I figure that one of the biggest impediments to this migration is actually going to be upgrading the PB/SB IDEs (unless, of course, these keep using v3 even if the Scintilla PureLib become v4).

This new Git repository could be the ideal grounds to test the Scintilla migration (both the PureLib _and_ its application in the PB IDE), by taking on Scintilla 4 in a beta branch until the IDE sources have been fully upgraded and the new PureLib thoroughly tested. Thanks to Git, it's possible to this on dedicated branches, while at the same time being able to integrate other changes from `master` — but it would be preferable to leave any Scintilla related code untouched in `master` branch while the migration is under work, to prevent odd conflicts and headaches.

See also my post on the PB Forum (Jan, 2018):

- [Scintilla Update to v3.7.6 (before migrating to v4)]

[LongTerm3]: https://sourceforge.net/p/scintilla/code/ci/LongTerm3/tree/
[Scintilla Update to v3.7.6 (before migrating to v4)]: https://www.purebasic.fr/english/viewtopic.php?f=3&t=70071

**Comment by fantaisie-software** [comment]
I checked the requirement for the v4, and it needs MSVC++ 2017 and its redistributable, so it's a no go for us as we don't want to have such a dependencies for PB apps (it means than every app which will use the scintilla gadget will have to install the MSVC 2017 redistributable package to be sure the libs are installed).

**Comment by tajmone** [comment]
> I checked the requirement for the v4, and it needs MSVC++ 2017 and its redistributable

While it was still in Beta, I've managed to compile Scintilla 4 under Windows using an old version of GCC, both as DLL and static library. But didn't then really do much experimenting with it beyond compiling it without error. I remember that it wouldn't compile with GCC 6.3.0, so I had to download an older version (the recommended version for Linux by Scintilla) and it compiled without errors nor warnings — but then, this was a long time ago.

> so it's a no go for us as we don't want to have such a dependencies for PB apps

Not even as an alternative option to using the current Scintilla? I'm sure some users are willing to embrace the dependencies for the benefits of the new features of v4 — also, MSVC++ 2017 was already superseded by v2019, so it's not exactly a bleeding edge dependency either. 

Since we're speaking of Win dependencies, is PB still bound to MSVC 2010 when it comes to creating PureLibs?
